### PR TITLE
Feature: add meson cross-build toolchain

### LINKED ISSUE #4529: Meson and cross compilation
The meson build helper does not handle cross compilation in Conan 1.12.2.
According to http://mesonbuild.com/Cross-compilation.html it would require writing a cross build definition file and giving it as  --cross-file argument to meson at configuration time

- [X] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've specified the Conan version, operating system version and any tool that can be relevant.
- [ ] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

### PR DESCRIPTION
**Pull request by SSE4**
closes: #4529

/cc @ericLemanissier @TheQwertiest  @KristianJerpetjon 

generates meson cross-file
tested on Android NDK and iOS so far

Changelog: Feature: Add meson cross-build toolchain.
Docs: https://github.com/conan-io/docs/pull/1980

- [x] Refer to the issue that supports this Pull Request.
- [x] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [x] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [x] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

### CODE REVIEW

File: conans/test/integration/toolchains/android/test_using_meson.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
+import os
```
Comment (memsharded) [inline conans/test/integration/toolchains/android/test_using_meson.py:1]: Maybe we want to group toolchains tests by toolchain, so ``toolchains/meson/test_android.py`` and ``toolchains/meson/test_ios.py`` etc.

File: conan/tools/meson/meson.py [hunk 2]
```diff
@@ -1,6 +1,7 @@
 import os
 
 from conans.client.build.cppstd_flags import cppstd_from_settings
+from conans.client.tools.oss import cross_building, get_cross_building_settings
```
Comment (memsharded) [inline conan/tools/meson/meson.py:4]: We need to move this to the new context and profile-build, plus move needed tools to the new ``conan.tool`` namespace. But it is good, definitely not in this PR.

**Review by memsharded — APPROVED** [review]
This looks fantastic job.
The most important thing is to note that these tools are experimental, and that they might break in future releases.

I vote for releasing it asap in next 1.32 so we can put it in user hands for easier and fast feedback. Please don't use it yet in ConanCenter, I would say to wait unless 1 or 2 releases to make sure we can stabilize it.

File: conan/tools/meson/meson.py [hunk 3]
```diff
@@ -159,21 +179,95 @@ def _native_content(self):
...
+        cpu_tuple = cpu_family_map.get(machine_arch, default_cpu_tuple)
+        cpu_family, cpu, endian = cpu_tuple[0], cpu_tuple[1], cpu_tuple[2]
+        context = {
+            'system': self._to_meson_value(system),
+            'cpu_family': self._to_meson_value(cpu_family),
+            'cpu': self._to_meson_value(cpu),
+            'endian': self._to_meson_value(endian),
+        }
+        t = Template(self._machine_template)
+        return t.render(context)
```
Comment (jgsogo) [inline conan/tools/meson/meson.py]:
```suggestion
        return self._render(self._machine_template, context)
```

Let's use that _helper_ function always (or never)

File: conan/tools/meson/meson.py [hunk 4]
```diff
@@ -159,21 +179,95 @@ def _native_content(self):
...
 
     @property
     def _cross_content(self):
-        raise Exception("cross-building is not implemented yet!")
+        os_build, arch_build, os_host, arch_host = get_cross_building_settings(self._conanfile)
+        os_target, arch_target = os_host, arch_host  # assume target the same as a host for now?
```
Comment (jgsogo) [inline conan/tools/meson/meson.py]: If we can ignore the `target_machine` maybe it is better (Conan doesn't implement it), if not, at least add a TODO/FIXME

```suggestion
        os_target, arch_target = os_host, arch_host  # TODO: assume target the same as a host for now?
```

File: conans/test/integration/toolchains/android/test_using_meson.py [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+import platform
+import textwrap
+import unittest
+
+from parameterized import parameterized
+
+from conans.test.assets.sources import gen_function_cpp, gen_function_h
+from conans.test.utils.tools import TestClient
+
+from ..test_meson import get_meson_version
```
Comment (jgsogo) [inline conans/test/integration/toolchains/android/test_using_meson.py]:
```suggestion
from conans.test.integration.toolchains.test_meson import get_meson_version
```

Why not use `Meson.get_version()` static method?

File: conans/test/integration/toolchains/ios/test_using_meson.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+import textwrap
+import unittest
+
+from parameterized import parameterized
+
+from conans.client.tools.apple import XCRun, apple_deployment_target_flag, to_apple_arch
+from conans.test.assets.sources import gen_function_cpp, gen_function_h
+from conans.test.utils.tools import TestClient
+
+from ..test_meson import get_meson_version
```
Comment (jgsogo) [inline conans/test/integration/toolchains/ios/test_using_meson.py]:
```suggestion
from conans.test.integration.toolchains.test_meson import get_meson_version
```

**Review by jgsogo — APPROVED** [review]
🙌

**Pull request merged**